### PR TITLE
Compression of conditional formats.

### PR DESCRIPTION
**Pull request by b0bi79**
Now, any manipulations with ranges lead to the breaking of the conditional formats. If we have a large amount of data covered in conventional formats, then any manipulation with the region becomes terribly long and consumes a large amount of memory. The Compress method merges the conditional formats.

### CODE REVIEW

**Comment by igitur** [comment]
Thanks.

Do you have an large example file that illustrates the memory usage problem?

**Comment by igitur** [comment]
PS: Sign up at https://hacktoberfest.digitalocean.com/ , submit 3 more pull requests and you get a free T-shirt.

**Comment by b0bi79** [comment]
Here is a test that shows the issue.
[Subranges_Simple_tMD1.xlsx](https://github.com/ClosedXML/ClosedXML/files/1354865/Subranges_Simple_tMD1.xlsx)

```
        [Fact]
        public void Test()
        {
            var fileName = Path.Combine(TestConstants.GaugesFolder, "Subranges_Simple_tMD1.xlsx");
            var workbook = new XLWorkbook(fileName);
            var srcSheet = workbook.Worksheet(1);

            srcSheet.Row(8).InsertRowsAbove(1);

            var dstSheet = workbook.AddWorksheet("Copy");

            for (int i = 0; i < 147; i++)
            {
                var srcRng = srcSheet.Range(i * 10 + 1, 1, i * 10 + 10, 9);
                var dstRng = dstSheet.Cell(i * 10 + 1, 1);
                srcRng.CopyTo(dstRng);
            }
            workbook.SaveAs(Path.Combine(TestConstants.GaugesFolder, "bigcopy.xlsx"));
        }
```

**Comment by igitur** [comment]
Does this replace #343?

**Comment by igitur** [comment]
And I ran your changes on that big file you attached. It doesn't look like it compresses the conditional formats. They are still broken up.

**Comment by b0bi79** [comment]
This is one part of that big PR #343.
I'm sorry, I had to give an explanation.
I attached the file to demonstrate the problem. In this file, formatting is initially compressed, and the test demonstrates the problem without compression.
Formats are fragmented from 200 to 65000. For this reason, the subsequent copying of the ranges becomes slow. A test on my computer runs for 25 seconds. And if compression is on, then 10 sec.
I have another series of improvements that were included in PR # 343. With these improvements, Compres will work more efficiently.
Here is an uncompressed [file](https://github.com/ClosedXML/ClosedXML/files/1358897/bigcopy.xlsx)

**Comment by b0bi79** [comment]
And yet, I found errors when copying the range. In some cases, formulas in conditional formatting are distorted.

**Comment by igitur** [comment]
Continue on this PR if possible.

**Comment by b0bi79** [comment]
Fixed

**Comment by b0bi79** [comment]
Intersection from XLRangeBase is cool, but very slow.

File: ClosedXML/Excel/ConditionalFormats/IXLConditionalFormats.cs [hunk 1]
```diff
@@ -10,5 +10,6 @@ public interface IXLConditionalFormats: IEnumerable<IXLConditionalFormat>
         void Add(IXLConditionalFormat conditionalFormat);
         void RemoveAll();
         void Remove(Predicate<IXLConditionalFormat> predicate);
+        void Compress();
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/IXLConditionalFormats.cs]: 1. Ok, I will fix it.
2. Excel automatically consolidates the conditional formats. I can make it private, but how then to test this method?
3. Ok.

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs [hunk 2]
```diff
@@ -1,11 +1,80 @@
...
+
+                var xxValues = xx.Values.Values.Where(v => !v.IsFormula).Select(v=>v.Value);
+                var yyValues = yy.Values.Values.Where(v => !v.IsFormula).Select(v => v.Value);
+                var xxFormulas = xx.Values.Values.Where(v => v.IsFormula).Select(f => ((XLCell)x.Range.FirstCell()).GetFormulaR1C1(f.Value));
+                var yyFormulas = yy.Values.Values.Where(v => v.IsFormula).Select(f => ((XLCell)y.Range.FirstCell()).GetFormulaR1C1(f.Value));
+
+                var xStyle = xx._style ?? xx.Range.Worksheet.Workbook.GetStyleById(xx._styleCacheId);
+                var yStyle = yy._style ?? yy.Range.Worksheet.Workbook.GetStyleById(yy._styleCacheId);
+
+                return Equals(xStyle, yStyle)
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs]: Ok

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs [hunk 3]
```diff
@@ -1,11 +1,80 @@
...
+                    && _listComparer.Equals(xxFormulas, yyFormulas)
+                    && _colorsComparer.Equals(xx.Colors, yy.Colors)
+                    && _contentsTypeComparer.Equals(xx.ContentTypes, yy.ContentTypes)
+                    && _iconSetTypeComparer.Equals(xx.IconSetOperators, yy.IconSetOperators)
+                    && (!_compareRange || Equals(xx.Range.RangeAddress, yy.Range.RangeAddress)) ;
+            }
+
+            public int GetHashCode(IXLConditionalFormat obj)
+            {
+                throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs]: Ok

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs [hunk 4]
```diff
@@ -13,6 +13,22 @@ public void Add(IXLConditionalFormat conditionalFormat)
             _conditionalFormats.Add(conditionalFormat);
         }
 
+        private bool RangeAbove(IXLRangeAddress newAddr, IXLRangeAddress addr)
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs]: Ok

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs [hunk 5]
```diff
@@ -13,6 +13,22 @@ public void Add(IXLConditionalFormat conditionalFormat)
...
+        {
+            return newAddr.FirstAddress.ColumnNumber == addr.FirstAddress.ColumnNumber
+                   && newAddr.LastAddress.ColumnNumber == addr.LastAddress.ColumnNumber
+                   && newAddr.FirstAddress.RowNumber < addr.FirstAddress.RowNumber
+                   && (newAddr.LastAddress.RowNumber+1).Between(addr.FirstAddress.RowNumber, addr.LastAddress.RowNumber);
+        }
+
+        private bool RangeBefore(IXLRangeAddress newAddr, IXLRangeAddress addr)
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs]: Ok

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs [hunk 6]
```diff
@@ -29,6 +45,52 @@ public void Remove(Predicate<IXLConditionalFormat> predicate)
...
+                        Merge(bottom, format);
+                        _conditionalFormats.Remove(format);
+                    }
+                }
+            }
+        }
+
+        private static void Merge(IXLConditionalFormat format, IXLConditionalFormat item)
+        {
+            foreach (var v in format.Values.ToList())
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs]: ToList removed

File: ClosedXML/Excel/ConditionalFormats/IXLConditionalFormats.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (igitur) [reply ClosedXML/Excel/ConditionalFormats/IXLConditionalFormats.cs]: If you make it `internal` it will be visible to the tests project.

**Comment by igitur** [comment]
Some more questions, should conditional formats automatically be consolidated (better word than compressed) when a file is loaded? Maybe it should be an option. Same for the saving... maybe we should add it to `SaveOptions`.? It depends on how much additional runtime it takes. The attachment in #579 is another example of a file with a LOT of conditional formats.

File: ClosedXML/Excel/ConditionalFormats/IXLConditionalFormats.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/IXLConditionalFormats.cs]: Thanks

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs [hunk 7]
```diff
@@ -29,6 +45,52 @@ public void Remove(Predicate<IXLConditionalFormat> predicate)
             _conditionalFormats.RemoveAll(predicate);
         }
 
+        public void Compress()
+        {
+            var formats = _conditionalFormats
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs]: I could not think of the conditions under which method Consolidate should throw an exception. This method consolidates CFs that can be consolidated and does not change CFs that should not be changed.

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs [hunk 8]
```diff
@@ -29,6 +45,52 @@ public void Remove(Predicate<IXLConditionalFormat> predicate)
...
+                .OrderByDescending(x=>x.Range.RangeAddress.FirstAddress.RowNumber)
+                .ThenByDescending(x=>x.Range.RangeAddress.FirstAddress.ColumnNumber)
+                .ToList();
```
[the remark this answers is not in the document]
Comment (b0bi79) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormats.cs]: Ok, ToList removed

File: ClosedXML/Excel/XLWorkbook.cs [hunk 9]
```diff
@@ -487,6 +487,7 @@ public void SaveAs(String file, Boolean validate, Boolean evaluateFormulae = fal
         public void SaveAs(String file, SaveOptions options)
         {
             checkForWorksheetsPresent();
+
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Remove this line that you accidentally added.

File: ClosedXML/Excel/XLWorkbook.cs [hunk 10]
```diff
@@ -575,6 +576,7 @@ public void SaveAs(Stream stream, Boolean validate, Boolean evaluateFormulae = f
         public void SaveAs(Stream stream, SaveOptions options)
         {
             checkForWorksheetsPresent();
+
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook.cs]: Remove this line that you accidentally added.

File: ClosedXML_Tests/ClosedXML_Tests.csproj [hunk 11]
```diff
@@ -87,6 +87,7 @@
     <Compile Include="Excel\CalcEngine\StatisticalTests.cs" />
     <Compile Include="Excel\CalcEngine\TextTests.cs" />
     <Compile Include="Excel\Comments\CommentsTests.cs" />
+    <Compile Include="Excel\ConditionalFormats\ConditionalFormatesCompressTests.cs" />
```
Comment (igitur) [inline ClosedXML_Tests/ClosedXML_Tests.csproj]: Rename to `ConditionalFormatsConsolidateTests.cs`

File: ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs [hunk 12]
```diff
@@ -0,0 +1,136 @@
+﻿using System.Linq;
+using ClosedXML.Excel;
+using NUnit.Framework;
+
+namespace ClosedXML_Tests.Excel.ConditionalFormats
+{
+    [TestFixture]
+    public class ConditionalFormatsCompressTests
```
Comment (igitur) [inline ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs]: Rename to `ConditionalFormatsConsolidateTests`

File: ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        [Test]
+        public void ConsecutivelyRowsCompressTest()
```
Comment (igitur) [inline ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs]: Rename to `ConsecutivelyRowsConsolidateTest`

File: ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+            ((XLConditionalFormats)ws.ConditionalFormats).Consolidate();
+
+            Assert.AreEqual(1, ws.ConditionalFormats.Count());
+            var format = ws.ConditionalFormats.First();
+            Assert.AreEqual("B2:C4", format.Range.RangeAddress.ToStringRelative());
+            Assert.AreEqual("F2", format.Values.Values.First().Value);
+        }
+
+        [Test]
+        public void ConsecutivelyColumnsCompressTest()
```
Comment (igitur) [inline ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs]: Rename to `ConsecutivelyColumnsConsolidateTest`

File: ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            SetFormat1(ws.Range("B8:B8").AddConditionalFormat());
+            SetFormat1(ws.Range("B9:C9").AddConditionalFormat());
+
+            ((XLConditionalFormats)ws.ConditionalFormats).Consolidate();
+
+            Assert.AreEqual(3, ws.ConditionalFormats.Count());
+        }
+
+        [Test]
+        public void DifferentFormatNoCompressTest()
```
Comment (igitur) [inline ClosedXML_Tests/Excel/ConditionalFormats/ConditionalFormatesCompressTests.cs]: Replace in all test names `Compress` with `Consolidate`.

**Comment by igitur** [comment]
Just a few more minor changes, please.

**Comment by b0bi79** [comment]
Done

**Comment by igitur** [comment]
So only one thing keeping me from merging this. I'm trying to determine which is best:
consolidate automatically on load/save or make the method public to the user and let him choose when to consolidate. For spreadsheets with lots of CF rules, automatic consolidation will make the load/save very slow. I know I told you to make the method internal, but now I'm having my doubts.

What is your opinion?

**Comment by b0bi79** [comment]
I optimized the consolidation code, it works quite quickly. 65000 CFs compresses in 1/3 of a second. I think that saves will take a little less time. You can check the speed of the consolidation.

**Pull request merged**